sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -406,11 +406,14 @@ object ScalaReflection extends ScalaReflection {
     def toCatalystArray(input: Expression, elementType: `Type`): Expression = {
       val externalDataType = dataTypeFor(elementType)
       val Schema(catalystType, nullable) = silentSchemaFor(elementType)
-      if (isNativeType(catalystType)) {
-        NewInstance(
-          classOf[GenericArrayData],
-          input :: Nil,
-          dataType = ArrayType(catalystType, nullable))
+      if (isNativeType(externalDataType)) {
+        expressions.If(
+          IsNull(input),
+          expressions.Literal.create(null, ArrayType(catalystType, nullable)),
+          NewInstance(
+            classOf[GenericArrayData],
+            input :: Nil,
+            dataType = ArrayType(catalystType, nullable)))
```

Comment:
The part of fixing is from pr #10401. Without this, the following fixing to `MapObjects` will cause null exception. Can we merge #10401 first then I do rebase here?